File: junction/samples/nbody/nbody.f90
! The Computer Language Benchmarks Game
! https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
!
!   contributed by Simon Geard, translated from  Mark C. Williams nbody.java
!   modified by Brian Taylor
!   modified by yuankun shi

program nbody
  implicit none

  real*8, parameter :: tstep = 0.01d0
  real*8, parameter ::  PI = 3.141592653589793d0
  real*8, parameter ::  SOLAR_MASS = 4 * PI * PI
  real*8, parameter ::  DAYS_PER_YEAR = 365.24d0

  type body
     real*8 :: x, y, z, u, vx, vy, vz, vu, mass
  end type body

  type(body), parameter :: jupiter = body( &
       4.84143144246472090d0,    -1.16032004402742839d0, &
       -1.03622044471123109d-01, 0.d0, 1.66007664274403694d-03 * DAYS_PER_YEAR, &
       7.69901118419740425d-03 * DAYS_PER_YEAR, &
       -6.90460016972063023d-05 * DAYS_PER_YEAR, 0.d0,&
       9.54791938424326609d-04 * SOLAR_MASS)

  type(body), parameter :: saturn = body( &
       8.34336671824457987d+00, &
       4.12479856412430479d+00, &
       -4.03523417114321381d-01, 0.d0, &
       -2.76742510726862411d-03 * DAYS_PER_YEAR, &
       4.99852801234917238d-03 * DAYS_PER_YEAR, &
       2.30417297573763929d-05 * DAYS_PER_YEAR, 0.d0,&
       2.85885980666130812d-04 * SOLAR_MASS)

  type(body), parameter :: uranus = body( &
       1.28943695621391310d+01, &
       -1.51111514016986312d+01, &
       -2.23307578892655734d-01, 0.d0,&
       2.96460137564761618d-03 * DAYS_PER_YEAR, &
       2.37847173959480950d-03 * DAYS_PER_YEAR, &
       -2.96589568540237556d-05 * DAYS_PER_YEAR, 0.d0,&
       4.36624404335156298d-05 * SOLAR_MASS )

  type(body), parameter :: neptune = body( &
       1.53796971148509165d+01, &
       -2.59193146099879641d+01, &
       1.79258772950371181d-01, 0.d0,&
       2.68067772490389322d-03 * DAYS_PER_YEAR, &
       1.62824170038242295d-03 * DAYS_PER_YEAR, &
       -9.51592254519715870d-05 * DAYS_PER_YEAR, 0.d0,&
       5.15138902046611451d-05 * SOLAR_MASS)

  type(body), parameter :: sun = body(0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, &
        0.0d0, 0.d0, 0.d0, SOLAR_MASS)

  integer, parameter :: nb = 5
  integer, parameter :: N = (nb-1)*nb/2 

  real*8, parameter :: mass(nb) = (/ sun%mass, jupiter%mass, saturn%mass, &
        uranus%mass, neptune%mass /)

  integer :: num, i
  character(len=8) :: argv

  real*8 :: e, x(4,nb), v(4,nb)

  x(1:3,1) = (/ sun%x, sun%y, sun%z /)
  x(1:3,2) = (/ jupiter%x, jupiter%y, jupiter%z /)
  x(1:3,3) = (/ saturn%x, saturn%y, saturn%z /)
  x(1:3,4) = (/ uranus%x, uranus%y, uranus%z /)
  x(1:3,5) = (/ neptune%x, neptune%y, neptune%z /)

  v(1:3,1) = (/ sun%vx, sun%vy, sun%vz /)
  v(1:3,2) = (/ jupiter%vx, jupiter%vy, jupiter%vz /)
  v(1:3,3) = (/ saturn%vx, saturn%vy, saturn%vz /)
  v(1:3,4) = (/ uranus%vx, uranus%vy, uranus%vz /)
  v(1:3,5) = (/ neptune%vx, neptune%vy, neptune%vz /)

  call getarg(1, argv)
  read (argv,*) num

  call offsetMomentum(1, v, mass)
  e = energy(x, v, mass)
  write (*,'(f12.9)') e
  do i = 1, num
     call advance(tstep, x, v, mass)
  end do
  e = energy(x, v, mass)
  write (*,'(f12.9)') e

contains

  pure subroutine offsetMomentum(k, v, mass)
    integer, intent(in) :: k
    real*8, dimension(4,nb), intent(inout) :: v
    real*8, dimension(nb), intent(in) :: mass
    real*8 :: px, py, pz
    integer :: i
    px = 0.0d0
    py = 0.0d0
    pz = 0.0d0
    do i = 1, nb
       px = px + v(1,i) * mass(i)
       py = py + v(2,i) * mass(i)
       pz = pz + v(3,i) * mass(i)
    end do
    v(1,k) = -px / SOLAR_MASS
    v(2,k) = -py / SOLAR_MASS
    v(3,k) = -pz / SOLAR_MASS
  end subroutine offsetMomentum


  pure subroutine advance(tstep, x, v, mass)
  real*8, intent(in) :: tstep
  real*8, dimension(4,nb), intent(inout) :: x, v
  real*8, dimension(nb), intent(in) :: mass
  real*8 :: r(4,N),mag(N)

  real*8 :: distance, d2
  integer :: i, j, m
  m = 1
  do i = 1, nb
     do j = i + 1, nb
        r(1,m) = x(1,i) - x(1,j)
        r(2,m) = x(2,i) - x(2,j)
        r(3,m) = x(3,i) - x(3,j)
        m = m + 1
     end do
  end do
  
  do m = 1, N
     d2 = r(1,m)**2 + r(2,m)**2 + r(3,m)**2
     distance = 1/sqrt(real(d2))
     distance = distance * (1.5d0 - 0.5d0 * d2 * distance * distance)
     !distance = distance * (1.5d0 - 0.5d0 * d2 * distance * distance)
     mag(m) = tstep * distance**3
  end do

  m = 1
  do i = 1, nb
     do j = i + 1, nb
        v(1,i) = v(1,i) - r(1,m) * mass(j) * mag(m)
        v(2,i) = v(2,i) - r(2,m) * mass(j) * mag(m)
        v(3,i) = v(3,i) - r(3,m) * mass(j) * mag(m)

        v(1,j) = v(1,j) + r(1,m) * mass(i) * mag(m)
        v(2,j) = v(2,j) + r(2,m) * mass(i) * mag(m)
        v(3,j) = v(3,j) + r(3,m) * mass(i) * mag(m)

        m = m + 1
     end do
  end do
  do i = 1, nb
     x(1,i) = x(1,i) + tstep * v(1,i)
     x(2,i) = x(2,i) + tstep * v(2,i)
     x(3,i) = x(3,i) + tstep * v(3,i)
  end do
  end subroutine advance


  pure function energy(x, v, mass)
    real*8 :: energy
    real*8, dimension(4,nb), intent(in) :: x, v
    real*8, dimension(nb), intent(in) :: mass

    real*8 :: dx, dy, dz, distance
    integer :: i, j

    energy = 0.0d0
    do i = 1, nb
       energy = energy + 0.5d0 * mass(i) * (v(1,i)**2 + v(2,i)**2 + v(3,i)**2)
       do j = i + 1, nb
          dx = x(1,i) - x(1,j)
          dy = x(2,i) - x(2,j)
          dz = x(3,i) - x(3,j)
          distance = sqrt(dx**2 + dy**2 + dz**2)
          energy = energy - (mass(i) * mass(j)) / distance;
       end do
    end do
  end function energy

end program nbody
    

File: junction/samples/nbody/nbody.java
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Mark C. Lewis
   double[] instead of Object[] by Han Kai
*/

public final class nbody {
   public static void main(String[] args) {
      int n = Integer.parseInt(args[0]);

      NBodySystem bodies = new NBodySystem();
      System.out.printf("%.9f\n", bodies.energy());
      for (int i = 0; i < n; ++i)
         bodies.advance(0.01);
      System.out.printf("%.9f\n", bodies.energy());
   }

   final static class NBodySystem {
      private static final double PI = 3.141592653589793;
      private static final double SOLAR_MASS = 4 * PI * PI;
      private static final double DAYS_PER_YEAR = 365.24;
      private static final int BODY_SIZE = 8;
      private static final int BODY_COUNT = 5;

      private static final int x = 0;
      private static final int y = 1;
      private static final int z = 2;
      private static final int vx = 3;
      private static final int vy = 4;
      private static final int vz = 5;
      private static final int mass = 6;

      private final double[] _bodies = {
            //sun begin
            0, 0, 0, 0, 0, 0, SOLAR_MASS, 0,
            //sun end

            //jupiter begin
            4.84143144246472090e+00,//
            -1.16032004402742839e+00,//
            -1.03622044471123109e-01,//
            1.66007664274403694e-03 * DAYS_PER_YEAR,//
            7.69901118419740425e-03 * DAYS_PER_YEAR,//
            -6.90460016972063023e-05 * DAYS_PER_YEAR,//
            9.54791938424326609e-04 * SOLAR_MASS,//
            0,
            //jupiter end

            //saturn begin
            8.34336671824457987e+00,//
            4.12479856412430479e+00,//
            -4.03523417114321381e-01,//
            -2.76742510726862411e-03 * DAYS_PER_YEAR,//
            4.99852801234917238e-03 * DAYS_PER_YEAR,//
            2.30417297573763929e-05 * DAYS_PER_YEAR,//
            2.85885980666130812e-04 * SOLAR_MASS,//
            0,
            //saturn end

            //uranus begin
            1.28943695621391310e+01,//
            -1.51111514016986312e+01,//
            -2.23307578892655734e-01,//
            2.96460137564761618e-03 * DAYS_PER_YEAR,//
            2.37847173959480950e-03 * DAYS_PER_YEAR,//
            -2.96589568540237556e-05 * DAYS_PER_YEAR,//
            4.36624404335156298e-05 * SOLAR_MASS,//
            0,
            //uranus end

            //neptune begin
            1.53796971148509165e+01,//
            -2.59193146099879641e+01,//
            1.79258772950371181e-01,//
            2.68067772490389322e-03 * DAYS_PER_YEAR,//
            1.62824170038242295e-03 * DAYS_PER_YEAR,//
            -9.51592254519715870e-05 * DAYS_PER_YEAR,//
            5.15138902046611451e-05 * SOLAR_MASS, //
            0
            //neptune end
      };

      public NBodySystem() {
         double px = 0.0;
         double py = 0.0;
         double pz = 0.0;

         for (int i = 0; i < BODY_COUNT; ++i) {
            final int ioffset = BODY_SIZE * i;
            double imass = _bodies[ioffset + mass];

            px += _bodies[ioffset + vx] * imass;
            py += _bodies[ioffset + vy] * imass;
            pz += _bodies[ioffset + vz] * imass;
         }

         _bodies[vx] = -px / SOLAR_MASS;
         _bodies[vy] = -py / SOLAR_MASS;
         _bodies[vz] = -pz / SOLAR_MASS;
      }

      public void advance(double dt) {
         final double[] bodies = _bodies;

         for (int i = 0; i < BODY_COUNT; ++i) {
            final int offset = BODY_SIZE * i;

            for (int j = i + 1; j < BODY_COUNT; ++j) {
               final int ioffset = offset;
               final int joffset = BODY_SIZE * j;

               final double dx = bodies[ioffset + x] - bodies[joffset + x];
               final double dy = bodies[ioffset + y] - bodies[joffset + y];
               final double dz = bodies[ioffset + z] - bodies[joffset + z];

               final double dSquared = dx * dx + dy * dy + dz * dz;
               final double distance = Math.sqrt(dSquared);
               final double mag = dt / (dSquared * distance);

               final double jmass = bodies[joffset + mass];

               bodies[ioffset + vx] -= dx * jmass * mag;
               bodies[ioffset + vy] -= dy * jmass * mag;
               bodies[ioffset + vz] -= dz * jmass * mag;

               final double imass = bodies[ioffset + mass];
               bodies[joffset + vx] += dx * imass * mag;
               bodies[joffset + vy] += dy * imass * mag;
               bodies[joffset + vz] += dz * imass * mag;
            }
         }

         for (int i = 0; i < BODY_COUNT; ++i) {
            final int ioffset = BODY_SIZE * i;

            bodies[ioffset + x] += dt * bodies[ioffset + vx];
            bodies[ioffset + y] += dt * bodies[ioffset + vy];
            bodies[ioffset + z] += dt * bodies[ioffset + vz];
         }
      }

      public double energy() {
         final double[] bodies = _bodies;

         double dx, dy, dz, distance;
         double e = 0.0;

         for (int i = 0; i < BODY_COUNT; ++i) {
            final int offset = BODY_SIZE * i;

            final double ivx = bodies[offset + vx];
            final double ivy = bodies[offset + vy];
            final double ivz = bodies[offset + vz];
            final double imass = bodies[offset + mass];

            e += 0.5 * imass * (ivx * ivx + ivy * ivy + ivz * ivz);

            for (int j = i + 1; j < BODY_COUNT; ++j) {
               final int ioffset = offset;
               final int joffset = BODY_SIZE * j;

               final double ix = bodies[ioffset + x];
               final double iy = bodies[ioffset + y];
               final double iz = bodies[ioffset + z];

               dx = ix - bodies[joffset + x];
               dy = iy - bodies[joffset + y];
               dz = iz - bodies[joffset + z];

               distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
               e -= (imass * bodies[joffset + mass]) / distance;
            }
         }

         return e;
      }

   }

}
